Validate being called on structs that are nil

In the case that the `reflect.Value` is `nil` even if it implements the `Validator` interface `Validate` should not be called. Before 0.8.0 `Validate` would not be called, causing issues with existing beats and libbeat.

One example of an error in libbeat:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x43c8945]

goroutine 44 [running]:
testing.tRunner.func1(0xc00016e500)
        /usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x451d380, 0x4a41c80)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/elastic/beats/libbeat/common/transport/tlscommon.(*CertificateConfig).Validate(...)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/libbeat/common/transport/tlscommon/types.go:256
github.com/elastic/beats/libbeat/common/transport/tlscommon.(*Config).Validate(0x0, 0x4522fa0, 0x0)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/libbeat/common/transport/tlscommon/config.go:97 +0x5
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.tryValidate(0x4522fa0, 0xc00046a290, 0x196, 0x4522fa0, 0xc00046a290)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/validator.go:123 +0xf3
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.tryRecursiveValidate(0x4522fa0, 0xc00046a290, 0x196, 0xc00016c400, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/validator.go:163 +0x132
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.reifyGetField(0xc00015a900, 0xc00016c400, 0x0, 0x0, 0x0, 0x0, 0x44ade00, 0x3, 0x4522fa0, 0xc00046a290, ...)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/reify.go:317 +0x19a
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.reifyStruct(0xc00016c400, 0x4594f00, 0xc00046a1b0, 0x199, 0xc00015a900, 0x0, 0x0)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/reify.go:282 +0x939
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.reifyInto(0xc00016c400, 0x44df7a0, 0xc00046a1b0, 0x16, 0xc00015a900, 0x0, 0x4594f00)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/reify.go:172 +0x3fb
github.com/elastic/beats/vendor/github.com/elastic/go-ucfg.(*Config).Unpack(0xc00015a900, 0x44df7a0, 0xc00046a1b0, 0x4a4ea60, 0x3, 0x3, 0x24, 0xc000064e90)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/vendor/github.com/elastic/go-ucfg/reify.go:155 +0x103
github.com/elastic/beats/libbeat/common.(*Config).Unpack(...)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/libbeat/common/config.go:168
github.com/elastic/beats/libbeat/outputs/logstash.readConfig(0xc00015a900, 0x45be607, 0x8, 0x45bd0f3, 0x3, 0x0, 0x0, 0x45bd108, 0x3, 0x0, ...)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/libbeat/outputs/logstash/config.go:78 +0x171
github.com/elastic/beats/libbeat/outputs/logstash.TestConfig.func1(0xc00016e500)
        /Users/blake/Code/elastic/beats/src/github.com/elastic/beats/libbeat/outputs/logstash/config_test.go:90 +0x89
testing.tRunner(0xc00016e500, 0xc00046a120)
        /usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x350
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Removed
 
 ### Fixed
+- Prevent Validate from being called when value is a pointer or interface and is nil. #144
 
 ## [0.8.0]
 

--- a/validator.go
+++ b/validator.go
@@ -151,6 +151,11 @@ func tryRecursiveValidate(val reflect.Value, opts *options, validators []validat
 		return nil
 	}
 
+	t := val.Type()
+	if (t.Kind() == reflect.Ptr || t.Kind() == reflect.Interface) && val.IsNil() {
+		return nil
+	}
+
 	var err error
 	switch chaseValue(val).Kind() {
 	case reflect.Struct:

--- a/validator.go
+++ b/validator.go
@@ -110,6 +110,10 @@ func tryValidate(val reflect.Value) error {
 	t := val.Type()
 	var validator Validator
 
+	if (t.Kind() == reflect.Ptr || t.Kind() == reflect.Interface) && val.IsNil() {
+		return nil
+	}
+
 	if t.Implements(tValidator) {
 		validator = val.Interface().(Validator)
 	} else if reflect.PtrTo(t).Implements(tValidator) {

--- a/validator_test.go
+++ b/validator_test.go
@@ -50,6 +50,16 @@ type structWithValidationTags struct {
 	I int `validate:"positive"`
 }
 
+type nestedPtrStructValidator struct {
+	A *ptrStructValidator `validate:"required"`
+	B *ptrStructValidator `validate:"required"`
+}
+
+type nestedNestedPtrStructValidator struct {
+	A *nestedPtrStructValidator
+	B *nestedPtrStructValidator
+}
+
 var errZeroTest = errors.New("value must not be 0")
 var errEmptyTest = errors.New("value must not be empty")
 var errMoreTest = errors.New("value must have more than 1 element")
@@ -284,6 +294,11 @@ func TestValidationPass(t *testing.T) {
 		&struct {
 			X *ptrStructValidator // Validator not called as its nil value
 		}{},
+		&struct {
+			X *nestedNestedPtrStructValidator
+		}{
+			X: &nestedNestedPtrStructValidator{},
+		},
 	}
 
 	for i, test := range tests {
@@ -504,6 +519,11 @@ func TestValidationFail(t *testing.T) {
 		&struct {
 			X int `validate:"required"`
 		}{},
+		&struct {
+			X *nestedPtrStructValidator
+		}{
+			X: &nestedPtrStructValidator{},
+		},
 	}
 
 	for i, test := range tests {

--- a/validator_test.go
+++ b/validator_test.go
@@ -281,6 +281,9 @@ func TestValidationPass(t *testing.T) {
 		&struct {
 			X int // field not present in config, but not required
 		}{},
+		&struct {
+			X *ptrStructValidator // Validator not called as its nil value
+		}{},
 	}
 
 	for i, test := range tests {

